config_utilities/src/factory.cpp

```diff
@@ -0,0 +1,205 @@
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ * -------------------------------------------------------------------------- */
+
+#include "config_utilities/factory.h"
+
+namespace config::internal {
+
+std::unique_ptr<ModuleRegistry> ModuleRegistry::s_instance_ = nullptr;
+
+std::string ModuleInfo::argumentString(const std::string& separator, const std::string& cap) const {
+  if (arguments.empty()) {
+    return "";
+  }
+
+  if (arguments.size() == 1 && skip_first_arg) {
+    return cap + "_" + cap;
+  }
+
+  auto iter = arguments.begin();
+  if (skip_first_arg) {
+    ++iter;
+  }
+
+  std::string arg_list = skip_first_arg ? "_" + separator : cap;
+  while (iter != arguments.end()) {
+    arg_list.append(*iter);
+    ++iter;
+    if (iter != arguments.end()) {
+      arg_list.append(separator);
+    }
+  }
+
+  return arg_list + cap;
+}
+
+std::string ModuleInfo::typeInfo() const {
+  return "BaseT='" + base_type + "' and ConstructorArguments={" + argumentString(", '", "'") + "}";
+}
+
+std::string ModuleInfo::signature() const { return base_type + "(" + argumentString() + ")"; }
+
+bool operator<(const ModuleInfo& lhs, const ModuleInfo& rhs) {
+  if (lhs.base_type == rhs.base_type) {
+    return lhs.arguments < rhs.arguments;
+  }
+
+  return lhs.base_type < rhs.base_type;
+}
+
+bool operator<(const ConfigPair& lhs, const ConfigPair& rhs) {
+  if (lhs.base_type == rhs.base_type) {
+    return lhs.config_type < rhs.config_type;
+  }
+
+  return lhs.base_type < rhs.base_type;
+}
+
+std::string ModuleRegistry::getAllRegistered() {
+  std::stringstream ss;
+  ss << "Modules registered to factories: {";
+  for (const auto& [key, types] : instance().type_registry) {
+    ss << "\n  " << key.signature() << ": {";
+    for (const auto& [type_name, derived_type] : types) {
+      ss << "\n    '" << type_name << "' (" << derived_type << "),";
+    }
+    ss << "\n  },";
+  }
+  ss << "\n}";
+  return ss.str();
+}
+
+bool ModuleRegistry::hasModule(const ModuleInfo& key, const std::string& type) {
+  const auto& modules = instance().modules;
+  const auto iter = modules.find(key);
+  if (iter == modules.end()) {
+    return false;
+  }
+
+  return iter->second->hasEntry(type);
+}
+
+void ModuleRegistry::removeModule(const ModuleInfo& key, const std::string& type) {
+  {  // module scope
```

Comment:
Is there a reason for these being scoped?

Reply:
I didn't feel like coming up with specific variable names for `iter` so the scope is to be able to use it twice